index/scorch/optimize.go

```diff
@@ -191,22 +188,16 @@ OUTER:
 				continue OUTER
 			}
 
-			itr, ok := tfr.iterators[i].(*zap.PostingsIterator)
+			itr, ok := tfr.iterators[i].(segment.OptimizablePostingsIterator)
 			if !ok {
-				// We optimize zap postings iterators only.
+				// We only optimize postings iterators that support this operation.
 				return nil, nil
 			}
 
 			// If the postings iterator is "1-hit" optimized, then we
 			// can perform several optimizations up-front here.
 			docNum1Hit, ok := itr.DocNum1Hit()
 			if ok {
-				if docNum1Hit == zap.DocNum1HitFinished {
-					// An empty docNum here means the entire AND is empty.
-					oTFR.iterators[i] = segment.AnEmptyPostingsIterator
```

Comment:
My assertion is that after calling `itr.DocNum1Hit()`, if `ok` is true, then `docNum1Hit` can never be `zap.DocNum1HitFinished`.  I believe this is true because of the implementation of `DocNum1Hit()` has this logic:

https://github.com/blevesearch/bleve/blob/master/index/scorch/segment/zap/posting.go#L790-L791

The only case where it returns true, it already ensured that `p.docNum1Hit != DocNum1HitFinished`